src/TestProblem/Hydro/CCSN/Flag_CCSN.cpp

```diff
@@ -107,36 +110,81 @@ bool Flag_Lightbulb( const int i, const int j, const int k, const int lv, const
 
 
 // TODO: fine-tune the criteria
 // (1) always refined to highest level in the region with r < 30 km
-   if ( r * UNIT_L < 3e6 )
+   if      ( R * UNIT_L < 3e6 )
       Flag = true;
 
-   else
-   {
-//    (2-a) density is larger than the threshold in Input__Flag_User
-      if ( Rho[k][j][i] < Threshold[0] )   return false;
+// (2) density is larger than the threshold in Input__Flag_User
+   else if ( Rho[k][j][i] < Threshold[0] )
+      Flag = false;
 
-//    (2-b) the cell width at son level (lv+1) is larger than the threshold
-      const double Min_CellWidth = r * CCSN_MaxRefine_RadFac;
-
-      Flag = ( 0.5 * dh ) > Min_CellWidth;
-   }
+// (3) check if the minimum angular resolution (in degree) is reached
+   if ( CCSN_Min_AngRes > 0.0  &&  !Flag )
+      if ( R * CCSN_Min_AngRes < dh * RAD2DEG )
+         Flag = true;
 
 
    return Flag;
 
 } // FUNCTION : Flag_Lightbulb
+
+
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  Flag_Region_CCSN
+// Description :  Check if the element (i,j,k) of the input patch is within the regions allowed to be refined
+//                for the CCSN test problem
+//
+// Note        :  1. Invoked by Flag_Check() using the function pointer "Flag_Region_Ptr",
+//                   which must be set by a test problem initializer
+//                2. Enabled by the runtime option "OPT__FLAG_REGION"
+//
+// Parameter   :  i,j,k       : Indices of the target element in the patch ptr[0][lv][PID]
+//                lv          : Refinement level of the target patch
+//                PID         : ID of the target patch
+//
+// Return      :  "true/false"  if the input cell "is/is not" within the region allowed for refinement
+//-------------------------------------------------------------------------------------------------------
+bool Flag_Region_CCSN( const int i, const int j, const int k, const int lv, const int PID )
+{
+
+   const double dh     = amr->dh[lv];
+   const double Pos[3] = { amr->patch[0][lv][PID]->EdgeL[0] + (i+0.5)*dh,
+                           amr->patch[0][lv][PID]->EdgeL[1] + (j+0.5)*dh,
+                           amr->patch[0][lv][PID]->EdgeL[2] + (k+0.5)*dh  };
+
+   bool Within = true;
+
+
+   const double Center[3] = { 0.5*amr->BoxSize[0], 0.5*amr->BoxSize[1], 0.5*amr->BoxSize[2] };
+   const double dR[3]     = { Pos[0]-Center[0], Pos[1]-Center[1], Pos[2]-Center[2] };
+   const double R         = sqrt( SQR(dR[0]) + SQR(dR[1]) + SQR(dR[2]) );
+
+// check minimum angular resolution (in degree) is reached
+   if ( CCSN_Min_AngRes > 0.0 )
+      if ( R * CCSN_Min_AngRes < dh * RAD2DEG )
+         return true;
+
+// check allowed maximum refine level based on angular resolution (in degree)
+   if ( CCSN_Max_AngRes > 0.0 )
+   {
+      if ( 2.0 * R * CCSN_Max_AngRes > dh * RAD2DEG )   Within = false;
+
+      if ( !Within )   return Within;
```

Comment:
- combine the statements of `CCSN_Min_AngRes` and `CCSN_Max_AngRes`. We do not need to check `CCSN_Min_AngRes` if `CCSN_Max_AngRes` is disabled.
- remove `if ( !Within )   return Within;` since there are no other criteria below currently.
- Is `(in degree)` necessary? I add that in my tentative code just for reminders

Reply:
- I am not so sure about this part, so I left it. Don't we need `CCSN_Min_AngRes` even `CCSN_Max_AngRes` is disabled?
- I have removed this.
- I removed this comment.

Reply:
No. `Flag_Region()` is used to check whether the input patch is allowed for refinement. 

If `Within = false`, we will not check if other refinement criteria (Lohner, `Flag_Corecollapse/Lightbulb`, etc) are satisfied. On the other hand, `Within = true` will not affect whether the input patch will be refined or not.

Reply:
I modified this part. Please check this.